### PR TITLE
Add Phi3 Mini 4K Instruct Model to torchtune

### PR DESCRIPTION
**Pull request by kartikayk**
Imp Note: The tokenizer still needs some work, this will be a follow up PR.

#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Changelog

This PR adds support for the Phi3 Mini 4K Instruct model to torchtune. Specifically we add the following:
- Phi3's RoPE module. This is not numerically equivalent to the Llama2 or Llama3 models and care needs to be taken to have correct behavior for bf16 training
- State dict conversion logic accounting for the fused qkv and gate_up projection matrices
- Config for multi-gpu full finetuning

#### Test plan
Please make sure to do each of the following if applicable to your PR. (If you're not sure about any one of these just ask and we will happily help.)

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [x] add unit tests for any new functionality
- [x] update docstrings for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [x] manually run any new or modified recipes with sufficient proof of correctness
	- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

Unit Tests

Detailed comparisons with [reference implementation](https://gist.github.com/kartikayk/87db294c9c9791622140cf4c9dc5a628)

```
pytest tests/torchtune
```

Full-finetune Recipe

Screenshot 2024-04-25 at 7 57 45 PM

### CODE REVIEW

File: recipes/configs/phi3/mini_full.yaml [hunk 1]
```diff
@@ -0,0 +1,78 @@
+# Config for multi-device full finetuning in full_finetune_distributed.py
+# using a Phi3 Mini 4K Instruct
+#
+# This config uses hyperparameters based on small set of experiments and information
```
Comment (joecummings) [inline recipes/configs/phi3/mini_full.yaml:4]: What experiments?

File: recipes/configs/phi3/mini_full.yaml [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+#   tune run --nproc_per_node 4 recipes/full_finetune_distributed.py --config recipes/configs/phi3/mini_full.yaml checkpointer.checkpoint_dir=<YOUR_CHECKPOINT_DIR>
+#
+# This config works best when the model is being fine-tuned on 2+ GPUs.
+# Single device full finetuning requires more memory optimizations. It's
+# best to use 7B_full_single_device.yaml for those cases
+
+# Tokenizer
+tokenizer:
+  _component_: torchtune.models.phi3.phi3_tokenizer
+  path: /data/users/kartikayk/cpts/Phi-3-mini-4k-instruct/tokenizer.model
```
Comment (joecummings) [inline recipes/configs/phi3/mini_full.yaml:27]: Update path

File: recipes/configs/phi3/mini_full.yaml [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+seed: null
+shuffle: True
+
+# Model Arguments
+model:
+  _component_: torchtune.models.phi3.phi3_mini
+
+checkpointer:
+  _component_: torchtune.utils.FullModelHFCheckpointer
+  checkpoint_dir: /data/users/kartikayk/cpts/Phi-3-mini-4k-instruct
```
Comment (joecummings) [inline recipes/configs/phi3/mini_full.yaml:42]: Change checkpoint

File: recipes/configs/phi3/mini_full.yaml [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+  checkpoint_files: [
+    model-00001-of-00002.safetensors,
+    model-00002-of-00002.safetensors
+  ]
+  recipe_checkpoint: null
+  output_dir: /data/users/kartikayk/cpts/Phi-3-mini-4k-instruct
```
Comment (joecummings) [inline recipes/configs/phi3/mini_full.yaml:48]: Change checkpoint

File: recipes/configs/phi3/mini_full.yaml [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+enable_activation_checkpointing: True
+
+# Reduced precision
+dtype: bf16
+
+# Logging
+metric_logger:
+  _component_: torchtune.utils.metric_logging.DiskLogger
+  log_dir: ${output_dir}
+output_dir: /tmp/Mistral-7B-v0.1/
```
Comment (joecummings) [inline recipes/configs/phi3/mini_full.yaml:77]: Change from Mistral

File: recipes/configs/phi3/mini_full_low_memory.yaml [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+# This config assumes that you've run the following command before launching
+# this run:
+#   tune download microsoft/Phi-3-mini-4k-instruct --output-dir ./Phi-3-mini-4k-instruct --hf-token <HF_TOKEN> --ignore-patterns ""
+#
+# The default config uses an optimizer from bitsandbytes. If you do not have it installed,
+# you can install it with
+#   pip install bitsandbytes
+#
+# To launch on a single device, run the following command from root:
+#   tune run full_finetune_single_device --config recipes/config/phi3/mini_full_low_memory.yaml
```
Comment (joecummings) [inline recipes/configs/phi3/mini_full_low_memory.yaml:17]:
```suggestion
#   tune run full_finetune_single_device --config phi3/mini_full_low_memory.yaml
```

File: recipes/configs/phi3/mini_full_low_memory.yaml [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+#
+# You can add specific overrides through the command line. For example
+# to override the checkpointer directory while launching training
+# you can run:
+#   tune run full_finetune_single_device --config recipes/config/phi3/mini_full_low_memory.yaml checkpointer.checkpoint_dir=<YOUR_CHECKPOINT_DIR>
```
Comment (joecummings) [inline recipes/configs/phi3/mini_full_low_memory.yaml:22]:
```suggestion
#   tune run full_finetune_single_device --config phi3/mini_full_low_memory.yaml checkpointer.checkpoint_dir=<YOUR_CHECKPOINT_DIR>
```

File: recipes/configs/phi3/mini_full_low_memory.yaml [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+#
+# This config works only for training on single device.
+
+# Tokenizer
+tokenizer:
+  _component_: torchtune.models.phi3.phi3_tokenizer
+  path: /data/users/kartikayk/cpts/Phi-3-mini-4k-instruct/tokenizer.model
```
Comment (joecummings) [inline recipes/configs/phi3/mini_full_low_memory.yaml:29]: Path

File: recipes/configs/phi3/mini_full_low_memory.yaml [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+seed: null
+shuffle: True
+
+# Model Arguments
+model:
+  _component_: torchtune.models.phi3.phi3_mini
+
+checkpointer:
+  _component_: torchtune.utils.FullModelHFCheckpointer
+  checkpoint_dir: /data/users/kartikayk/cpts/Phi-3-mini-4k-instruct
```
Comment (joecummings) [inline recipes/configs/phi3/mini_full_low_memory.yaml:44]: Path

File: recipes/configs/phi3/mini_full_low_memory.yaml [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+  checkpoint_files: [
+    model-00001-of-00002.safetensors,
+    model-00002-of-00002.safetensors
+  ]
+  recipe_checkpoint: null
+  output_dir: /data/users/kartikayk/cpts/Phi-3-mini-4k-instruct
```
Comment (joecummings) [inline recipes/configs/phi3/mini_full_low_memory.yaml:50]: Path

File: recipes/configs/phi3/mini_full_low_memory.yaml [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+dtype: bf16
+
+# Model compilation
+compile: False
+
+# Logging
+metric_logger:
+  _component_: torchtune.utils.metric_logging.DiskLogger
+  log_dir: ${output_dir}
+output_dir: /tmp/Mistral-7B-v0.1/
```
Comment (joecummings) [inline recipes/configs/phi3/mini_full_low_memory.yaml:82]: Path

File: tests/torchtune/modules/test_position_embeddings.py [hunk 12]
```diff
@@ -97,3 +98,45 @@ def test_rope_init_meta_device(self, input_params):
...
+        _, _, head_dim, _, max_seq_len = input_params
+        return Phi3RotaryPositionalEmbeddings(dim=head_dim, max_seq_len=max_seq_len)
+
+    def test_forward(
+        self, input: tensor, rope_phi3: Phi3RotaryPositionalEmbeddings
+    ) -> None:
+        x_out = rope_phi3(input)
+
+        # check the numerics of the computed tensor
+        assert_expected(x_out.mean(), tensor(-0.0005), atol=1e-4)
```
Comment (joecummings) [inline tests/torchtune/modules/test_position_embeddings.py:138]: Can you include in the PR where these numbers come from?

File: torchtune/models/phi3/_model_builders.py [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+        embed_dim=3072,
+        intermediate_dim=8192,
+        max_seq_len=4096,
+        attn_dropout=0.0,
+        norm_eps=1e-5,
+    )
+
+def phi3_tokenizer(path: str) -> SentencePieceTokenizer:
+    tokenizer = SentencePieceTokenizer(path)
+    tokenizer.pad_id = 32000
```
Comment (ebsmothers) [inline torchtune/models/phi3/_model_builders.py:41]: Looking at the [HF config](https://huggingface.co/microsoft/Phi-3-mini-4k-instruct/blob/main/config.json) it says that eos_id and pad_id are the same, is that right?

File: torchtune/models/phi3/_model_builders.py [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+from torchtune.modules.tokenizers import SentencePieceTokenizer
+from torchtune.modules.peft import LORA_ATTN_MODULES
+from functools import partial
+
+import torch
+
+
+"""
+Model builders build specific instantiations using component builders. For example
+the ``mistral_7b`` model builder uses the ``mistral`` component builder.
```
Comment (joecummings) [inline torchtune/models/phi3/_model_builders.py:15]: Update from mistral

File: torchtune/models/phi3/_position_embeddings.py [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+    """
+    RoPE Embeddings used in the Phi3 model.
+    Ref: https://huggingface.co/microsoft/Phi-3-mini-4k-instruct
+
+    This class is not numerically equivalent to the RoPE Embedding module
+    used by Llama2 and Llama3.
+
+    Args:
+        dim (int): Embedding dimension. This is usually set to the dim of each
+            head in the attention module computed as ````embed_dim`` // ``num_heads````
```
Comment (ebsmothers) [inline torchtune/models/phi3/_position_embeddings.py:24]: too many backticks?

File: torchtune/models/phi3/_position_embeddings.py [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+# LICENSE file in the root directory of this source tree.
+
+from typing import Optional
+
+import torch
+
+from torch import nn, Tensor
+
+
+class Phi3RotaryPositionalEmbeddings(nn.Module):
```
Comment (joecummings) [inline torchtune/models/phi3/_position_embeddings.py:14]: You're a hero

File: tests/torchtune/modules/test_position_embeddings.py (hunk 12 above)
Comment (kartikayk) [reply tests/torchtune/modules/test_position_embeddings.py:138]: Info's in the gist! Don't really want to replicate all of the info in the context again

File: torchtune/utils/_checkpointing/_checkpointer.py [hunk 17]
```diff
@@ -381,12 +382,15 @@ def load_checkpoint(self) -> Dict[str, Any]:
             del state_dict
             gc.collect()
 
-        converted_state_dict[utils.MODEL_KEY] = convert_weights.hf_to_tune(
-            merged_state_dict,
-            num_heads=self._config["num_attention_heads"],
-            num_kv_heads=self._config["num_key_value_heads"],
-            dim=self._config["hidden_size"],
-        )
+        if self._model_type == ModelType.PHI3_MINI:
```
Comment (joecummings) [inline torchtune/utils/_checkpointing/_checkpointer.py:385]: Does capitalization change this? B/c in the config it's `PHI_MINI` and in the enum it's `phi_mini`

**Review by joecummings — COMMENTED** [review]
Mostly nits! Thanks @kartikayk

File: torchtune/models/phi3/_position_embeddings.py [hunk 18]
```diff
@@ -0,0 +1,105 @@
...
+        idx_theta = torch.einsum("i, j -> ij", seq_idx, self.theta).float()
+
+        # We cache the cos and sin embeddings instead of the IDs. This helps
+        # ensure we have correct behavior when training with bf16
+        # Size: [max_seq_len, (dim * 2)]
+        freqs = torch.cat([idx_theta, idx_theta], dim=-1)
+        cache = torch.cat([freqs.cos(), freqs.sin()], dim=-1)
+        self.register_buffer("cache", cache, persistent=False)
+
+    def forward(self, x: Tensor, input_pos: Optional[Tensor] = None) -> Tensor:
```
Comment (ebsmothers) [inline torchtune/models/phi3/_position_embeddings.py:68]: input_pos is not used?

File: torchtune/models/phi3/_position_embeddings.py [hunk 19]
```diff
@@ -0,0 +1,105 @@
...
+        Notation used for tensor shapes:
+            - b: batch size
+            - s: sequence length
+            - n_h: num heads
+            - h_d: head dim
+
+        TODO: The implementation below can be made more efficient
+        for inference.
+        """
+        # input tensor has shape [b, s, n_h, n_d]
```
Comment (ebsmothers) [inline torchtune/models/phi3/_position_embeddings.py:88]: (based on the above docstring)
```suggestion
        # input tensor has shape [b, s, n_h, h_d]
```

File: recipes/configs/phi3/mini_full.yaml (hunk 1 above)
Comment (kartikayk) [reply recipes/configs/phi3/mini_full.yaml:4]: Copy paste!

File: recipes/configs/phi3/mini_full_low_memory.yaml (hunk 6 above)
Comment (kartikayk) [reply recipes/configs/phi3/mini_full_low_memory.yaml:17]: oh I havent added this to the registry, so this won't work. I'll let you take care of that

File: recipes/configs/phi3/mini_full_low_memory.yaml (hunk 7 above)
Comment (kartikayk) [reply recipes/configs/phi3/mini_full_low_memory.yaml:22]: same

File: torchtune/utils/_checkpointing/_checkpointer.py (hunk 17 above)
Comment (kartikayk) [reply torchtune/utils/_checkpointing/_checkpointer.py:385]: Yup, it matters!

File: torchtune/models/phi3/_model_builders.py (hunk 13 above)
Comment (kartikayk) [reply torchtune/models/phi3/_model_builders.py:41]: Yes, but also will let @joecummings confirm that!

File: torchtune/models/phi3/_position_embeddings.py (hunk 15 above)
Comment (kartikayk) [reply torchtune/models/phi3/_position_embeddings.py:24]: argh

File: torchtune/models/phi3/_position_embeddings.py (hunk 18 above)
Comment (kartikayk) [reply torchtune/models/phi3/_position_embeddings.py:68]: Fixed!

File: torchtune/models/phi3/_position_embeddings.py (hunk 19 above)
Comment (kartikayk) [reply torchtune/models/phi3/_position_embeddings.py:88]: oh good catch!

**Review by joecummings — APPROVED** [review]
Looks good!
